storage/src/index/mod.rs

```diff
@@ -14,1064 +14,1492 @@
         });
     }
 
     #[test_traced]
-    fn test_index_entry_replacement_not_a_collision() {
+    #[should_panic(expected = "must call Cursor::next()")]
+    fn test_ordered_index_update_before_next_panics() {
         let runner = deterministic::Runner::default();
         runner.start(|context| async move {
-            let mut index = Index::init(context.clone(), TwoCap);
+            let mut index = new_ordered(context);
+            run_index_update_before_next_panics(&mut index);
+        });
+    }
 
-            index.insert(b"a", 1); // collisions = 0
+    fn run_index_entry_replacement_not_a_collision<I: Index<Value = u64>>(index: &mut I) {
+        index.insert(b"a", 1);
+        {
             let mut cur = index.get_mut(b"a").unwrap();
-            cur.next(); // Entry
-            cur.delete(); // list empty, pruned = 1
-            cur.next(); // Done
-            cur.insert(2); // replacement, *not* collision
+            cur.next();
+            cur.delete();
+            cur.next();
+            cur.insert(2);
+        }
+        assert_eq!(index.keys(), 1);
+        assert_eq!(index.items(), 1);
+    }
 
-            assert!(context.encode().contains("keys 1"));
-            assert!(context.encode().contains("items 1"));
+    #[test_traced]
+    fn test_hash_index_entry_replacement_not_a_collision() {
+        let runner = deterministic::Runner::default();
+        runner.start(|context| async move {
+            let mut index = new_unordered(context);
+            run_index_entry_replacement_not_a_collision(&mut index);
         });
     }
 
     #[test_traced]
-    fn test_index_large_collision_chain_stack_overflow() {
+    fn test_ordered_index_entry_replacement_not_a_collision() {
         let runner = deterministic::Runner::default();
         runner.start(|context| async move {
-            let mut index = Index::init(context, TwoCap);
+            let mut index = new_ordered(context);
+            run_index_entry_replacement_not_a_collision(&mut index);
+        });
+    }
 
-            // Create a large collision chain with 50,000 items on the same key
-            for i in 0..50000 {
-                index.insert(b"", i as u64);
-            }
+    fn run_index_large_collision_chain_stack_overflow<I: Index<Value = u64>>(index: &mut I) {
+        for i in 0..50000 {
+            index.insert(b"", i as u64);
+        }
+    }
 
-            // Should not stack overflow (using default Rust drop, it will)
+    #[test_traced]
+    fn test_hash_index_large_collision_chain_stack_overflow() {
+        let runner = deterministic::Runner::default();
+        runner.start(|context| async move {
+            let mut index = new_unordered(context);
+            run_index_large_collision_chain_stack_overflow(&mut index);
         });
     }
 
     #[test_traced]
-    fn test_cursor_across_threads() {
+    fn test_ordered_index_large_collision_chain_stack_overflow() {
```

Comment:
I suppose it would make sense to macro this suite but this is consistent with `runtime` for now.